tiledb/sm/buffer/buffer.h

```diff
@@ -276,6 +303,148 @@ class Buffer {
+   * @param buff The buffer the object will encapsulate, working on its
+   *     data and size, but using a separate local offset, without affecting
+   *     the input buffer.
+   */
+  explicit ConstBuffer(Buffer* buff);
+
+  /** Destructor. */
+  ~ConstBuffer() = default;
+
+  /** Returns the buffer data. */
+  const void* data() const;
+
+  /** Returns pointer to data at the current offset. */
+  const void* cur_data() const;
+
+  /** Returns the number of bytes left for reading. */
+  uint64_t nbytes_left_to_read() const;
+
+  /**
+   * This is a special function used for reading from a buffer that stores
+   * uint64_t values. It reads a number of bytes from the local buffer and
+   * writes them to the input buffer, after adding *offset* to each read
+   * uint64_t value.
+   *
+   * @param buf The buffer to write to when reading from the local buffer.
+   * @param nbytes The number of bytes to read.
+   * @param offset The offset to add to each uint64_t value.
+   * @return void.
+   */
+  void read_with_shift(uint64_t* buf, uint64_t nbytes, uint64_t offset);
+
+  /**
+   * Returns a value from the buffer of type T.
+   *
+   * @tparam T The type of the value to be read.
+   * @param offset The offset in the local buffer to read from.
+   * @return The desired value of type T.
+   */
+  template <class T>
+  inline T value(uint64_t offset) const {
+    return ((const T*)(((const char*)data_) + offset))[0];
+  }
+
+  /**
+   * Returns the value at the current offset of the buffer of type T.
+   *
+   * @tparam T The type of the value to be returned.
+   * @return The value to be returned of type T.
+   */
+  template <class T>
+  inline T value() const {
+    return ((const T*)(((const char*)data_) + offset_))[0];
+  }
+};
+
+// =======================================================
+//      PreallocatedBuffer
+// =======================================================
+/**
+ * Writeable buffer that uses pre-allocated storage provided externally. Does
+ * not expand storage on write.
+ */
+class PreallocatedBuffer : public BufferBase {
+ public:
+  /** Default constructor deleted. */
+  PreallocatedBuffer() = delete;
+
+  /**
+   * Constructor.
+   *
+   * @param data The data of the buffer.
+   * @param size The size of the buffer.
+   */
+  PreallocatedBuffer(const void* data, uint64_t size);
+
+  /** Returns the buffer data pointer at the current offset. */
+  void* cur_data() const;
+
+  /** Returns the buffer data. */
+  const void* data() const;
+
+  /** Returns the "free space" in the buffer, which is the size minus the
+   * current offset. */
+  uint64_t free_space() const;
+
+  /**
+   * Returns a value from the buffer of type T.
+   *
+   * @tparam T The type of the value to be read.
+   * @param offset The offset in the local buffer to read from.
+   * @return The desired value of type T.
+   */
+  template <class T>
+  inline T value(uint64_t offset) const {
+    return ((const T*)(((const char*)data_) + offset))[0];
+  }
```

Comment:
A)BufferBase placement?
B)assert() for offset validity?

Reply:
Ditto.

Reply:
may be doable, see above comments.